includes/server/class-llms-rest-students-progress-controller.php

```diff
@@ -448,17 +450,48 @@ public function register_routes() {
 	 * Determine if current user has permission to create/update an enrollment.
 	 *
 	 * @since 1.0.0-beta.1
+	 * @since [version] Allow students to update their own progress for available posts.
+	 *                      Return 404s for non-existent and incompletable posts.
 	 *
 	 * @param WP_REST_Request $request Request object.
 	 * @return true|WP_Error
 	 */
 	public function update_item_permissions_check( $request ) {
 
-		if ( ! current_user_can( 'edit_post', $request['post_id'] ) || ! current_user_can( 'edit_students', $request['id'] ) ) {
-			return llms_rest_authorization_required_error();
+		$post = get_post( $request['post_id'] );
+		if ( ! $post ) {
+
+			// Post doesn't exist.
+			return llms_rest_not_found_error( __( 'The requested post resource could not be found.', 'lifterlms' ) );
+		} else if ( ! in_array( $post->post_type, llms_get_completable_post_types(), true ) ) {
+
+			// Post is not completable.
+			return llms_rest_not_found_error( __( 'The requested post is not completable.', 'lifterlms' ) );
+		} elseif ( current_user_can( 'edit_post', $request['post_id'] ) && current_user_can( 'edit_students', $request['id'] ) ) {
+
+			// Admin or manager who can mark things complete on behalf of the student.
+			return true;
```

Comment:
Ok those can update, but only existing progresses, I think.
So, either here or later we need a check on whether the progress exist, which. We have to defined what's an existing progress. To me, as _said_ in https://github.com/gocodebox/lifterlms-rest/issues/181#issuecomment-662944310 a progress exists if it exists in the db (whether the user is enrolled or not), or if the user is enrolled in the course/membership related to the `$request['post_id']` (whether the progress actually exists in the db or not). What do you think?

Reply:
During an update it shouldn't matter whether or not the progress exists, that should only matter when reading the object (for a GET request), right? So in this particular condition I don't think it matters whether any progress currently exists for the record, right?

Unless you're saying that if it's *already complete* and they try to complete it again it should return an error (or vice-versa if it's already incomplete and they try to mark it incomplete) but I'm not sure that matters. I think we can return the same response whether or not the progress has actually been changed, right? This would be (again I think) how REST is supposed to work. If you make the same request twice you should get the same response both times.

Reply:
Well I meant that whoever I am, if I want to `update` a progress of a student in a course I can do it only if this progress exists.
With my definition of "progress exists" above, I will be not able to update any progress which has no records in the db for unenrolled students.
Let me explain what I mean:
1) Student 1 completed Course 1 but then he was unenrolled
2) Student 2 is enrolled in Course 1 but he has no progress record in the db, because he never completed any lesson.
3) Student 3 is not enrolled in Course 1 and he never completed any lesson, there's no progress record in the db.

I must be able to update Student 1 p on Course 1, I guess, and I must be able to update Student 2 p on Course 1 as well, right? Although I must not be able to update a Student 3 on Course 1 because it's something that doesn't exist.

Does it make sense to you?

Reply:
Oh yes, I see what you're saying now:

1. Student 1 completed Course 1 but then he was unenrolled - **should** be able to update
2. Student 2 is enrolled in Course 1 but he has no progress record in the db, because he never completed any lesson -- **should** be able to update
3. Student 3 is not enrolled in Course 1 and he never completed any lesson, there's no progress record in the db -- **should not** be able to update

Is this what you're saying?

The admin condition in this method would allow 1 & 2 (which we want) and would also allow 3 (which we don't want).

I think we could actually just leave this completely alone and allow admins to perform condition 3 if they want to.

There's currently no UI in LifterLMS (on the admin panel) to accommodate this -- they would need to manually enroll first -- but adding an additional lookup to handle this condition feels unnecessary for the edge case. I think it's okay if the admin UI slightly differs from what's possible in this case.

What do you think?

If we agree that this is okay the method should be okay as is, right?

Reply:
Yes that's what I was saying.

Ok, yeah let's do as you propose then!